### DISCUSSION TITLE
Package `vite-plugin-svelte-docgen`

### DISCUSSION THREAD

**Discussion by xeho91** [svelte-docgen/svelte-docgen#16] [Ideas]
> [!IMPORTANT]
> We have a [consensus](https://github.com/svelte-docgen/svelte-docgen/discussions/16#discussioncomment-11559048)!
>

> **What should be the design direction of the Vite plugin for this package?**

There's several ways this plugin can work, but which one is demanded/expected to work by users?

## Current ideas

- [Transforming compiled output of *.svelte file](https://github.com/svelte-docgen/svelte-docgen/discussions/16#discussioncomment-11539931)
- [Creating virtual modules with `?docgen` import query suffix](https://github.com/svelte-docgen/svelte-docgen/discussions/16#discussioncomment-11539959)

**Comment by xeho91** [comment]
## Transforming compiled output of `*.svelte` file

This is how it works currently for [Storybook internal plugin](https://github.com/storybookjs/storybook/blob/next/code/frameworks/svelte-vite/src/plugins/svelte-docgen.ts).

### How it works?

For every `*.svelte` file, at the end of the compiled output. A property key `__docgen` with serialized docgen data is appended.

Example:

```js
export default function SomeComponentName() {
    // Compiled output logic
}
;__docgen = { /* serialized docgen data */ };
```

> [!NOTE]
> This example is based on Svelte `v5` - where the components are functions.
> In legacy version they're classes, but this doesn't change much, and the idea stays same.

This `SomeComponentName.__docgen` data is consumed later on internally, but that's a different subject I believe.

---


### Pros

?

### Cons

1. It affects the compiled output and can increase the build time
2. Depends on `@sveltejs/vite-plugin-svelte`, so the Vite plugin order will matter
3. Currently in the logic _(based on study of the Storybook internal plugin)_:
   - it's required to use path parsing with runtime I/O to read the original `*.svelte` file again,
     Because AFAIK  Vite plugin API doesn't have a way to forward original source file to the next plugin.
   - requires another preprocessing - ? _(I'm not sure yet as to why)_
   
   tl;dr I don't know how significantly, but it _potentially_ slows down the Vite build.

**Comment by xeho91** [comment]
## Creating virtual modules

I'm exploring idea for creating a virtual file _(module)_ with docgen data with Vite plugin.

### Example

```svelte
<!-- ./Counter.svelte -->
<script lang="ts">
	import type { HTMLButtonAttributes } from "svelte/elements";

	import docgen from "virtual:./Counter.docgen.js"; // 👈 Look at this, you can access this component type inside its own file

	interface Props extends HTMLButtonAttributes {}
	let { children, ...props }: Props = $props();
</script>

<button {...props}>
	{@render children?.()}
</button>

<details>
	<pre>
		{JSON.stringify(docgen)}
	</pre>
</details>
```

But in reality, users or tools will rather consume those docgen data outside the `*.svelte` component file. Regardless, it doesn't change anything. They can still be imported via:

```js
import docgen from "virtual:./Counter.docgen.js"; 
```

---


### Pros

1. Data can be generated **on-demand** - we don't have to generate for every existing `*.svelte` file
2. Unlike [transforming compiled output](https://github.com/svelte-docgen/svelte-docgen/discussions/16#discussioncomment-11539931) - it leaves it untouched. So, the file size doesn't increase.
3. Doesn't depend on the other plugins nor plugin order.

### Cons

1. Syntax can be a little confusing for users?
2. TypeScript experience? - still exploring 🔎

**Reply by xeho91** [reply]
I don't know if it is ever possible to let TypeScript be able to read virtual files generated by Vite. This would help infer **constants**. For example, let's assume the `Counter.svelte` file is a legacy component. 

```js
import docgen from "virtual:./Counter.docgen.js";

docgen.isLegacy;
//     ^ true
```

---

However, we can always declare modules via the `.d.ts` file, like so:

```ts
declare module "*.docgen.js" {
    isLegacy: boolean:
    // ...rest of schema
}
```

Following the aforementioned example, the type would look like this:

```js
import docgen from "virtual:./Counter.docgen.js";

docgen.isLegacy;
//     ^ boolean
```

Is still better than nothing!

**Reply by xeho91** [reply]
I've created a playground repository to explore this idea: https://github.com/svelte-docgen/vite-plugin-svelte-docgen-playground

**Reply by xeho91** [reply]
@ciscorn Suggested [another idea](https://bsky.app/profile/ciscorn.bsky.social/post/3ld5rrfywak2f) on the BlueSky.

It simplifies _(at least for me)_ import pattern into this:

```js
import componentDocgen from "./Component.svelte?docgen";
```


And, it's possible! 🎉 Commit with proof of concept: https://github.com/svelte-docgen/svelte-docgen/pull/19/commits/39e2522baa83f0a9e3313a9090815e0ada682071

**Comment by xeho91** [comment]
We have a consensus.

**We're going with virtual modules and using `?docgen` query in the module import path.**
